src/ir/block.h

```diff
@@ -0,0 +1,64 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_BLOCK_H_
+#define SRC_IR_BLOCK_H_
+
+#include <vector>
+
+#include "absl/container/flat_hash_map.h"
+#include "src/ir/operation.h"
+#include "src/ir/operator.h"
+#include "src/ir/types/type.h"
+#include "src/ir/value.h"
+
+namespace raksha::ir {
+
+// A collection of operations.
+class Block {
+ public:
+  // The type for a collection of `Operation` instances.
+  using OperationList = std::vector<std::unique_ptr<Operation>>;
+
+  Block() = default;
+
+  // Disable copy (and move) semantics.
+  Block(const Block&) = delete;
+  Block& operator=(const Block&) = delete;
+
+  const OperationList& operations() const { return operations_; }
+  const DataDeclMap& inputs() const { return inputs_; }
+  const DataDeclMap& outputs() const { return outputs_; }
+  const NamedValueListMap& results() const { return results_; }
+
+  friend class BlockBuilder;
+
+ private:
+  // The inputs to this block.
+  DataDeclMap inputs_;
+  // Outputs from this block.
+  DataDeclMap outputs_;
+  // The list of operations that belong to this block. This can be empty.
+  OperationList operations_;
+  // Maps the outputs of the operations in the list of operations in this block
+  // to the corresponding name. Note that a result can have more than one value
+  // which is used to represent non-determinism.
```

Comment:
If we had the `choose` operation, we could solve this without a result having more than one value by having the result be fed into with a `choose` with an unspecified condition.